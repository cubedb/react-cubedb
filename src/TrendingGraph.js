import * as d3 from 'd3'
import { filter, isUndefined, maxBy, sortBy, mapValues, toPairs } from 'lodash'
import React from 'react'
import PropTypes from 'prop-types'

import SvgLoadingAnimation from './utils/SvgLoadingAnimation'

import './style/TrendingGraph.scss'


const hourParseTime = d3.timeParse('%Y-%m-%d %H')
const dayParseTime = d3.timeParse('%Y-%m-%d')

const TRIM_LENGTH = 2


export default class TrendingGraph extends React.Component {
  static propTypes = {
    /**
     * Dictionary of dates and values
     */
    data: PropTypes.object,
    width: PropTypes.number,
    height: PropTypes.number,
    isLoading: PropTypes.bool
  }

  static defaultProps = {
    width: 200,
    height: 50
  }

  render() {
    const width = this.props.width
    const height = this.props.height
    let path

    if (this.props.isLoading) {
      path = <SvgLoadingAnimation width={width} height={height}/>
    } else if (this.props.data) {
      const data = sortBy(filter(toPairs(mapValues(this.props.data, 'c')), (val) => {
        return !isUndefined(val) && val.length === 2
      }), 0)
        .slice(TRIM_LENGTH, -TRIM_LENGTH)

      const maxValue = (maxBy(data, (e) => e[1])||[])[1]

      const parseTime = (dt) => (hourParseTime(dt)||dayParseTime(dt))

      const xAxis = d3.scaleTime()
        .rangeRound([0, width])
        .domain(d3.extent(data, (d) => parseTime(d[0])))

      const yAxis = d3.scaleLinear()
        .rangeRound([height*0.95, height*0.05])
        .domain([0, maxValue])

      const getLine = d3.line().x(d => xAxis(parseTime(d[0])))
        .y(d => yAxis(d[1]))

      const d = getLine(data)

      path = <path d={d} fill="none"
        strokeLinecap="round"
        strokeLinejoin="round"
        className="trending_graph__line"/>
    }

    return <svg className="trending_graph" viewBox={`0 0 ${width} ${height}`} width={width} height={height}>
      {path}
    </svg>
  }
}
