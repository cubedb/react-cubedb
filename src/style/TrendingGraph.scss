@import 'colors';

.trending_graph {
    display: inline-block;
    background: $background;
    border: 1px solid $border-soft;

    &__line {
        stroke: $primary;
    }
}
