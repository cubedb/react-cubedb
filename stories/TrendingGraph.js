import React from 'react'  // eslint-disable-line no-unused-vars

import { TrendingGraph } from '../src'
import { number } from '@storybook/addon-knobs'

const mock = require('./_mock.json')

export default () =>
  <div>
    <h2>Trending Graph</h2>
    <section>
      <h3>Default:</h3>
      <TrendingGraph data={mock.response.p}/>
    </section>
    <section>
      <h3>Without data:</h3>
      <TrendingGraph/>
    </section>
    <section>
      <h3>With custom dimension:</h3>
      <TrendingGraph
        data={mock.response.p}
        width={number('width', 200)}
        height={number('height',80)}/>
    </section>
  </div>
