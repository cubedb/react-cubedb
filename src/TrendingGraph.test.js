import React from "react"; // eslint-disable-line no-unused-vars
import { render } from "enzyme";

import TrendingGraph from "./TrendingGraph";

const defaultProps = {
  data: {
    1483232400: { c: 1 },
    1483236000: { c: 2 },
    1483239600: { c: 3 }
  }
};

describe("TrendingGraph", () => {
  it("default", () => {
    const component = render(<TrendingGraph {...defaultProps} />);

    expect(component).toMatchSnapshot();
  });

  it("no data", () => {
    const component = render(
      <TrendingGraph {...defaultProps} data={undefined} />
    );

    expect(component).toMatchSnapshot();
  });

  it("is loading", () => {
    const component = render(
      <TrendingGraph {...defaultProps} isLoading={true} />
    );

    expect(component).toMatchSnapshot();
  });
});
