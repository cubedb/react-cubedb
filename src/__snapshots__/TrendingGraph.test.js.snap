// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrendingGraph default 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <svg
    className="trending_graph"
    height={50}
    viewBox="0 0 200 50"
    width={200}
>
    <path
        className="trending_graph__line"
        d={null}
        fill="none"
        strokeLinecap="round"
        strokeLinejoin="round"
    />
</svg>,
  "nodes": Array [
    <svg
      className="trending_graph"
      height={50}
      viewBox="0 0 200 50"
      width={200}
>
      <path
            className="trending_graph__line"
            d={null}
            fill="none"
            strokeLinecap="round"
            strokeLinejoin="round"
      />
</svg>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance":  {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <TrendingGraph
        data={
                Object {
                        "2017-01-01 01": Object {
                          "c": 1,
                        },
                        "2017-01-01 02": Object {
                          "c": 2,
                        },
                        "2017-01-01 03": Object {
                          "c": 3,
                        },
                      }
        }
        height={50}
        width={200}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": TrendingGraph {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "data": Object {
            "2017-01-01 01": Object {
              "c": 1,
            },
            "2017-01-01 02": Object {
              "c": 2,
            },
            "2017-01-01 03": Object {
              "c": 3,
            },
          },
          "height": 50,
          "width": 200,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <svg
          className="trending_graph"
          height={50}
          viewBox="0 0 200 50"
          width={200}
>
          <path
                    className="trending_graph__line"
                    d={null}
                    fill="none"
                    strokeLinecap="round"
                    strokeLinejoin="round"
          />
</svg>,
        "_debugID": 2,
        "_renderedOutput": <svg
          className="trending_graph"
          height={50}
          viewBox="0 0 200 50"
          width={200}
>
          <path
                    className="trending_graph__line"
                    d={null}
                    fill="none"
                    strokeLinecap="round"
                    strokeLinejoin="round"
          />
</svg>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <TrendingGraph
    data={
        Object {
            "2017-01-01 01": Object {
              "c": 1,
            },
            "2017-01-01 02": Object {
              "c": 2,
            },
            "2017-01-01 03": Object {
              "c": 3,
            },
          }
    }
    height={50}
    width={200}
/>,
}
`;

exports[`TrendingGraph is loading 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <svg
    className="trending_graph"
    height={50}
    viewBox="0 0 200 50"
    width={200}
>
    <SvgLoadingAnimation
        height={50}
        width={200}
    />
</svg>,
  "nodes": Array [
    <svg
      className="trending_graph"
      height={50}
      viewBox="0 0 200 50"
      width={200}
>
      <SvgLoadingAnimation
            height={50}
            width={200}
      />
</svg>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance":  {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <TrendingGraph
        data={
                Object {
                        "2017-01-01 01": Object {
                          "c": 1,
                        },
                        "2017-01-01 02": Object {
                          "c": 2,
                        },
                        "2017-01-01 03": Object {
                          "c": 3,
                        },
                      }
        }
        height={50}
        isLoading={true}
        width={200}
/>,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": TrendingGraph {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "data": Object {
            "2017-01-01 01": Object {
              "c": 1,
            },
            "2017-01-01 02": Object {
              "c": 2,
            },
            "2017-01-01 03": Object {
              "c": 3,
            },
          },
          "height": 50,
          "isLoading": true,
          "width": 200,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <svg
          className="trending_graph"
          height={50}
          viewBox="0 0 200 50"
          width={200}
>
          <SvgLoadingAnimation
                    height={50}
                    width={200}
          />
</svg>,
        "_debugID": 6,
        "_renderedOutput": <svg
          className="trending_graph"
          height={50}
          viewBox="0 0 200 50"
          width={200}
>
          <SvgLoadingAnimation
                    height={50}
                    width={200}
          />
</svg>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <TrendingGraph
    data={
        Object {
            "2017-01-01 01": Object {
              "c": 1,
            },
            "2017-01-01 02": Object {
              "c": 2,
            },
            "2017-01-01 03": Object {
              "c": 3,
            },
          }
    }
    height={50}
    isLoading={true}
    width={200}
/>,
}
`;

exports[`TrendingGraph no data 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <svg
    className="trending_graph"
    height={50}
    viewBox="0 0 200 50"
    width={200}
/>,
  "nodes": Array [
    <svg
      className="trending_graph"
      height={50}
      viewBox="0 0 200 50"
      width={200}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance":  {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <TrendingGraph
        data={undefined}
        height={50}
        width={200}
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": TrendingGraph {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "data": undefined,
          "height": 50,
          "width": 200,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <svg
          className="trending_graph"
          height={50}
          viewBox="0 0 200 50"
          width={200}
/>,
        "_debugID": 4,
        "_renderedOutput": <svg
          className="trending_graph"
          height={50}
          viewBox="0 0 200 50"
          width={200}
/>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <TrendingGraph
    data={undefined}
    height={50}
    width={200}
/>,
}
`;
